shared/domains/chelonia/persistent-actions.js

```diff
@@ -0,0 +1,200 @@
+'use strict'
+
+import sbp from '@sbp/sbp'
+import { LOGIN } from '~/frontend/utils/events.js'
+
+type SbpInvocation = any[]
+
+type PersistentActionOptions = {
+  errorInvocation?: SbpInvocation,
+  // Maximum number of tries, default: Infinity.
+  maxAttempts: number,
+  // How many seconds to wait between retries.
+  retrySeconds: number,
+  skipCondition?: SbpInvocation,
+  // SBP selector to call on success with the received value.
+  successInvocationSelector?: string,
+  totalFailureInvocation?: SbpInvocation
+}
+
+type PersistentActionStatus = {
+  cancelled: boolean,
+  failedAttemptsSoFar: number,
+  lastError: string,
+  nextRetry: string,
+  pending: boolean
+}
+
+const defaultOptions: PersistentActionOptions = {
+  maxAttempts: Number.POSITIVE_INFINITY,
+  retrySeconds: 30
+}
+const tag = '[chelonia.persistentActions]'
+
+class PersistentAction {
+  invocation: SbpInvocation
+  options: PersistentActionOptions
+  status: PersistentActionStatus
+  timer: Object
+
+  constructor (invocation: SbpInvocation, options: PersistentActionOptions = {}) {
+    this.invocation = invocation
+    this.options = { ...defaultOptions, ...options }
+    this.status = {
+      cancelled: false,
+      failedAttemptsSoFar: 0,
+      lastError: '',
+      nextRetry: '',
+      pending: false
+    }
+  }
+
+  // Do not call if the action is pending or cancelled!
+  async attempt (): Promise<void> {
+    if (await this.trySBP(this.options.skipCondition)) {
+      this.cancel()
+      return
+    }
+    try {
+      this.status.pending = true
+      const result = await sbp(...this.invocation)
+      this.handleSuccess(result)
+    } catch (error) {
+      this.handleError(error)
+    }
+  }
+
+  cancel (): void {
+    this.timer && clearTimeout(this.timer)
+    this.status.cancelled = true
+    this.status.nextRetry = ''
+  }
+
+  async handleError (error: Error): Promise<void> {
+    const { options, status } = this
+    // Update relevant status fields before calling any optional selector.
+    status.failedAttemptsSoFar++
+    status.lastError = error.message
+    const anyAttemptLeft = options.maxAttempts > status.failedAttemptsSoFar
+    status.nextRetry = anyAttemptLeft && !status.cancelled
+      ? new Date(Date.now() + options.retrySeconds * 1e3).toISOString()
+      : ''
```

Comment:
Hmm, it doesn't seem like the value of `status.nextRetry` is actually used? 🤔

Reply:
That's because `setTimeout` isn't based on absolute time but on some relative number of milliseconds in the future. On the other hand, `status.nextRetry` is defined as an ISO timestamp, so it has to be absolute

Reply:
Right, but if it's not really serving a purpose then why have it at all?

The `if` condition below could be changed from `if (status.nextRetry)` to `if (anyAttemptLeft && !status.cancelled)`

Reply:
Oh, it's part of the description in the issue, that I wrote... lol. 🤦‍♂️